src/manage_sql_secinfo.c

```diff
@@ -2350,117 +2322,119 @@ insert_cve_products (element_t list, resource_t cve,
-                "    (SELECT id FROM scap2.cpes"
-                "     WHERE name='%s'))",
-                first_affected ? "" : ",", cve, quoted_product);
-            }
-          else
-            {
-              int cpe;
+          free (product_text);
+          product_element = element_next (product_element);
+          continue;
+        }
 
-              /* The product is in the db.
-               *
-               * So we don't need to insert it. */
+      product_decoded = g_uri_unescape_string (product_text, NULL);
+      product_tilde = string_replace (product_decoded,
+                                      "~", "%7E", "%7e",
+                                      NULL);
+      g_free (product_text);
+      g_free (product_decoded);
+      g_hash_table_insert (products, product_tilde, NULL);
 
-              cpe = hashed_cpes_cpe_id (hashed_cpes, product_tilde);
+      product_element = element_next (product_element);
+    }
+  
+  /* Add new CPEs. */
 
-              g_string_append_printf
-               (sql_affected,
-                "%s (%llu, %i)",
-                first_affected ? "" : ",", cve,
-                cpe);
-            }
+  first_product = first_affected = 1;
+  g_hash_table_iter_init (&products_iter, products);
 
-          first_affected = 0;
-          g_free (product_tilde);
+  while (g_hash_table_iter_next (&products_iter,
+                                 (gpointer*)(&product_tilde),
+                                 NULL))
+    {
+      if (g_hash_table_contains (hashed_cpes, product_tilde) == 0)
+        {
+          gchar *quoted_product;
+          quoted_product = sql_quote (product_tilde);
+          g_string_append_printf
+            (sql_cpes,
+             "%s ('%s', '%s', %i, %i)",
+             first_product ? "" : ",", quoted_product, quoted_product,
+             time_published, time_modified);
           g_free (quoted_product);
+          first_product = 0;
         }
+    }
 
-      g_free (product_text);
-
-      product = element_next (product);
+  if (first_product == 0)
+    {
+      /* Run the SQL for inserting new CPEs and add them to hashed_cpes 
+       * so they can be looked up quickly when adding affected_products.
+       */
+      iterator_t inserted_cpes;
+      init_iterator (&inserted_cpes,
+                     "%s"
+                     " ON CONFLICT (uuid)"
+                     " DO UPDATE SET name = EXCLUDED.name"
+                     " RETURNING id, name;",
+                     sql_cpes->str);
+
+      while (next (&inserted_cpes))
+        {
+          int rowid = iterator_int (&inserted_cpes, 0);
+          const char *name = iterator_string (&inserted_cpes, 1);
+          g_hash_table_insert (hashed_cpes,
+                               g_strdup (name),
+                               GINT_TO_POINTER (rowid));
+        }
+      cleanup_iterator (&inserted_cpes);
+      increment_transaction_size (transaction_size);
+      g_string_free (sql_cpes, TRUE);
     }
```

Comment:
Could freeing `sql_cpes` be moved outside the if-condition ?

```suggestion
      increment_transaction_size (transaction_size);
    }
    g_string_free (sql_cpes, TRUE);
```